app/jobs/csv_export_job.rb

```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
     filename = csv_export.path_file
     filter = csv_export.filters
 
-    @deploys = Deploy.joins(:stage, :job).where(filter)
+    get_deploys(filter)
     summary = ["-", "Generated At", csv_export.updated_at, "Deploys", @deploys.count.to_s]
     filters_applied = ["-", "Filters", filter.to_json]
 
     CSV.open(filename, 'w+') do |csv|
-      csv << [
-        "Deploy Number", "Project Name", "Deploy Sumary", "Deploy Commit", "Deploy Status", "Deploy Updated",
-        "Deploy Created", "Deployer Name", "Deployer Email", "Buddy Name", "Buddy Email",
-        "Production Flag", "No code deployed"
-      ]
+      csv << Deploy.csv_header
       @deploys.find_each do |deploy|
-        csv << [
-          deploy.id, deploy.project.name, deploy.summary, deploy.commit, deploy.job.status, deploy.updated_at,
-          deploy.start_time, deploy.job.user.name, deploy.job.user.try(:email), deploy.buddy_name, deploy.buddy_email,
-          deploy.stage.production, deploy.stage.no_code_deployed
-        ]
+        csv << deploy.csv_line
       end
       csv << summary
       csv << filters_applied
     end
   end
 
+  def get_deploys(filter)
+    if filter.keys.include?('environments.production')
+      production_value = filter.delete('environments.production')
+      # To match logic of stages.production? True when any deploy_group environment is true or
+      # deploy_groups environment is empty and stages is true
+      production_query = "(A.production = ? OR (A.production IS NULL AND stages.production = ?))"
+      production_query = "NOT " + production_query if production_value == false
+
+      # The query could result in duplicate entries when a stage has a production and non-production deploy group
+      # so it is important this is run only if enviornments.production was set
+      @deploys = Deploy.includes(:buddy, job: :user, stage_with_deleted: :project).
+        joins("INNER JOIN jobs ON jobs.id = deploys.job_id INNER JOIN stages ON stages.id = deploys.stage_id").
+        joins("LEFT JOIN " \
+          "(SELECT DISTINCT deploy_groups_stages.stage_id, environments.production FROM deploy_groups_stages " \
+          "INNER JOIN deploy_groups ON deploy_groups.id = deploy_groups_stages.deploy_group_id " \
+          "INNER JOIN environments ON environments.id = deploy_groups.environment_id) A ON A.stage_id = stages.id").
+        unscope(where: :deleted_at).where(filter).where(production_query, true, true)
+    else
+      @deploys = Deploy.includes(:buddy, job: :user, stage_with_deleted: :project).
```

Comment:
With this I was getting the expected output report, but had a performance improvement in generation time 10s for 20,000 deploys vs 28 seconds without.